Feature/scene tree properties

WeirdRubberDuck opened the pull request:
Adds implementation of the different property type views, as well as some styling for the menus with property owners.

**Needs testing!** 
There are definitely things that I have missed, so making this a draft PR. A test page can be accessed under "Windows", but note that this one does not cover all variants of the properties. 

Note that the built version runs much smoother than the dev one.

ylvaselling commented:
Overall I think it looks awesome!! ✨ I just jotted down some notes from looking at the properties. Lmk if you want to discuss 

1. General 
    1. Maybe we can make the titles less bold? They are a little hard to read now imo.

    1. I just had a thought - we could potientially go around the min max problem by making them as tiny input fields so you could set them yourself.

1. Numeric property 

    1. Overlap of text over the textbox 
![image](https://github.com/user-attachments/assets/fe319eeb-524c-4362-b0b0-37d0ac8120e2)

    1. Is there a way to make these look more connected? When the window is small it is hard to see that these two are the same property. 

    1. Why does the slider have dots for these values? Seems 0.05 is always in there?

    1. Should we somehow let the user know when they are out of range? Like we did in the old UI 

    1. Seems the circle slider doesn't go all the way to the right? I think there might be an offset to the blue bar? ![image](https://github.com/user-attachments/assets/bc19cc48-27aa-4592-b29e-d510c7e7592c)
![image](https://github.com/user-attachments/assets/67022d3d-cc22-4a02-9134-df68e7c633d5)

    1.  It seems when the property is line wrapped it goes outside bounds? ![image](https://github.com/user-attachments/assets/79a9278b-f890-407e-bd4b-5d61408806e7)

2. String list property
Can we somehow make this smaller? Like a dropdown?
![image](https://github.com/user-attachments/assets/2bbeba08-7243-4d9a-b25c-82bbb4e69da9)

8. Vec2 property
In the gui it took me a while to realize this was probably a vec2 property. I assume this is x and y? Can we maybe write that out somehow if that is the case? This might go for all vec properties - we might want to write out xyzw?
![image](https://github.com/user-attachments/assets/71fec6b3-bd9c-4f66-8745-dab44413eafe)

9. Selection property
Make a placeholder when empty?

4. Min max
Should we place the slider in the middle of the the number? Or the slider above the numbers, to mimic the numeric property when it is linewrapped?

3. Infobox
Can we place it so it is not in different places depending on how long the title is? Maybe to the far right?

5. Propertyowner
    1. Some padding on the right hand side? ![image](https://github.com/user-attachments/assets/53ae469c-7218-4d07-95d0-96e6fe8dd8d8)
    2. Align about with the upper border of the about text?
![image](https://github.com/user-attachments/assets/dadc5f01-1430-4b14-8480-0eff93ccc5db)

7. Seems we have an issue with rendering
![image](https://github.com/user-attachments/assets/d4e27fc5-51c7-484d-bab1-93f24544ec16)

6. Wth 😆 We really like saying Earth is a special place
![image](https://github.com/user-attachments/assets/a028d1f9-7ce6-465e-be0a-a4bbf2075f8f)

ylvaselling commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  const inputFieldRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    setShownValues(value);
+  }, [value]);
```
Use the hook you made for this?

ylvaselling commented on src/components/Property/Types/NumericProperty/NumericProperty.tsx:
```diff
@@ -0,0 +1,84 @@
+
+  // @TODO There still seems to be a bit of a stutter when dragging the slider.
+  // nvestigate. Is there somethign we could memo to make it better?
+  function onValueChange(newValue: number) {
```
onValueEnd? or something (update the value when releasing the slider)

WeirdRubberDuck replied:
Unfortunately, I donät think that is an option.  The user probably want to see the value bein updated while the slider is being used

ylvaselling commented on src/components/Property/Types/NumericProperty/NumericPropertySlider.tsx:
```diff
@@ -0,0 +1,115 @@
+  const scale = scalePow().exponent(exponent).domain([min, max]).range([min, max]);
+  const decimalPlaces = Math.max(0, -Math.floor(Math.log10(step)));
+
+  function markLabel(value: number, largeValuePrecision?: number): JSX.Element | string {
```
Make this a component?

ylvaselling commented on src/components/Property/Types/VectorProperty/VectorProperty.tsx:
```diff
@@ -20,11 +23,24 @@ export interface VectorPropertyProps {
-  if (value.length === 2 && viewOptions.MinMaxRange) {
-    return <MinMaxRange {...props} />;
+  let view;
+  if (isColor) {
+    view = <ColorVector {...props} />;
+  } else if (isMinMaxRange) {
+    view = <MinMaxRange {...props} />;
+  } else {
+    view = <ValueList {...props} />;
   }
 
-  return <ValueList {...props} />;
+  return (
+    <>
+      <PropertyLabel label={name} tip={description} />
+      {view}
+    </>
+  );
```
Make this a switch that uses viewOptions and make each case return?

WeirdRubberDuck replied:
Can't make a switch, but I'll make if statements for each view option, that directly returns

ylvaselling commented on src/components/Property/Types/VectorProperty/VectorProperty.tsx:
```diff
@@ -20,11 +23,24 @@ export interface VectorPropertyProps {
-  const { value, viewOptions } = props;
+  const { value, viewOptions, name, description } = props;
+
+  const isColor = (value.length === 3 || value.length === 4) && viewOptions.Color;
+  const isMinMaxRange = value.length === 2 && viewOptions.MinMaxRange;
```
Why do we need to check for this here? Can't we use only viewOption as the determining factor and throw an error in the component (ColorVector, MinMaxRange) if it is wrong? As it is now we allow the property to be set wrongly in openspace

WeirdRubberDuck replied:
This is simply copy-pasted from how it was implemented in the old UI, where it was simply a safeguard against if someone sets the wrong viewoption in C++ (which in general, we shouldn't), and cummincate in some way that this is the use case when each view option is supported on one line, I guess 😅  

But I agree, now that there are separate components for each view option, it would be nicer to check it in each component and throw of the number of values is wrong.  We should get away from the old tendency of silently ignoring errors

ylvaselling commented on src/components/Property/Types/VectorProperty/ViewOptions/ColorVector.tsx:
```diff
@@ -0,0 +1,38 @@
+
+import { ValueList } from './DefaultValueList';
+
+export function ColorVector(props: VectorPropertyProps) {
+  const { disabled, setPropertyValue, value } = props;
```
```suggestion
export function ColorVector({ disabled, setPropertyValue, value }: VectorPropertyProps) {

```

WeirdRubberDuck replied:
Here I would prefer to keep the props object, to make it easy to spread to the value list view

The props handling for the vector types properties might need some discussion, in general

WeirdRubberDuck replied:
@engbergandreas also commented on this. Let's discuss it :)

ylvaselling commented on src/components/Property/Types/VectorProperty/ViewOptions/ColorVector.tsx:
```diff
@@ -0,0 +1,38 @@
+  const { disabled, setPropertyValue, value } = props;
+  const hasAlpha = value.length === 4;
+
+  function setValue(newValue: number[]) {
+    setPropertyValue(newValue);
+  }
```
I don't think we need this function

ylvaselling commented on src/components/Property/Types/VectorProperty/ViewOptions/DefaultValueList.tsx:
```diff
@@ -16,28 +13,29 @@ export function ValueList({
   const max = additionalData.MaximumValue;
   const step = additionalData.SteppingValue;
 
+  const isInt = step.every((s) => s >= 1);
```
?

WeirdRubberDuck replied:
😅 A bit of a hacky way to check if the property type is an int... I'll make a new type instead, similar to the ListProperty

ylvaselling commented on src/components/Property/Types/NumericProperty/NumericProperty.tsx:
```diff
@@ -0,0 +1,84 @@
+
+import { NumericPropertySlider } from './NumericPropertySlider';
+
+interface Props {
+  name: string;
+  description: string;
+  disabled: boolean;
+  setPropertyValue: (newValue: number) => void;
+  value: number;
+  additionalData: {
+    Exponent: number;
+    MaximumValue: number;
+    MinimumValue: number;
+    SteppingValue: number;
+  };
+  // TODO: view options in metadata
+}
```
Can we extend Props from a Property type?

WeirdRubberDuck replied:
I would have to create a new type, but yes, it is possible. 

However, the only things I can really put in that type are `name`, `description`, and `disabled`. All the other arguments will be specific to the given property type

WeirdRubberDuck replied:
Would be good to discuss ideas for the property handling, if the current approach is confusing.

WeirdRubberDuck commented on src/components/PropertyOwner/PropertyOwner.tsx:
```diff
@@ -42,7 +42,9 @@ export function PropertyOwner({ uri, expandedOnDefault = false }: Props) {
       noTransition
     >
-      <PropertyOwnerContent uri={uri} />
+      <Box style={{ borderLeft: '2px solid var(--mantine-color-gray-8)' }}>
```
Investigate using mantine's Divider for this instead

WeirdRubberDuck commented on src/components/ColorPicker/ColorEdit.tsx:
```diff
@@ -0,0 +1,88 @@
+    <Stack gap={'xs'}>
+      <MantineColorPicker
+        value={value}
+        // We don't use the selected format here, since changin the format would then
```
```suggestion
        // We don't use the selected format here, since changing the format would then
```

WeirdRubberDuck commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,169 @@
+  disabled,
+  setPropertyValue,
+  value,
+  valueType = 'string'
```
It isn't. The default value was likely just left as a remnant from some old version.. :)

WeirdRubberDuck replied:
Actually, nvm.  It needs to be there. 

The int number and float number versions of the list properties is just a specialization of the regular string-based one (i.e. this one).

ylvaselling replied:
I would prefer it if there was one "base class" and then three "derived classes", i.e. make the StringListProperty component it's own component. It is not clear to me right now why the string is default 🤔 and if you have one specialized class for each property type you can place the specialized functions in them instead of having the switches

WeirdRubberDuck replied:
👍 that is a motivation to write it like that then. I'll change it

WeirdRubberDuck commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,169 @@
+  }, [value]);
+
+  // This needs to be a memoized function because it is used in a useEffect
+  const createPlaceHolderText = useCallback(
+    (values: ListValueType[]) => {
+      const count1 = values.length + 1;
+      const count2 = values.length + 2;
+      switch (valueType) {
+        case 'float':
+          return `number${count1}, number${count2}, ...`;
+        case 'int':
+          return `integer${count1}, integer${count2}, ...`;
+        case 'string':
+          return `item${count1}, item${count2}, ...`;
+        default:
+          throw new Error('Invalid value type');
+      }
+    },
+    [valueType]
+  );
+
```
Hehe, yeah, maybe this placeholder is a bit overdesigned... But I did find it a useful way to communicate what type of value the list input expects. There is currently on way for the user to know this, otherwise. 

Also since the inputs also do the work of just showing what the value is, I only added the placeholder on edit. Otherwise I think it would be unnecessarily cluttered

But I'll remove the count, seems redundant

engbergandreas commented on src/components/Property/Types/NumericProperty/NumericProperty.tsx:
```diff
@@ -0,0 +1,88 @@
+  // When no min/max is set, the marks for the slider cannot be nicely computed
+  const extent = max - min;
+  // @TODO (202503-03, emmbr) This should be handled a better way...
+  const maxAllowedExtentForSlider = 10e12;
```
Should it be 1e12 or 1e13 instead? 10e12 looks weird to me or am I forgetting something about scientific notations

engbergandreas commented on src/components/Property/Types/NumericProperty/Slider/NumericPropertySlider.tsx:
```diff
@@ -0,0 +1,71 @@
+import { usePropListeningState } from '@/api/hooks';
+import { useSliderScale } from '@/components/Property/SliderUtil/hook';
+
+import { SliderMinMaxLabels } from '../../../SliderUtil/SliderMinMaxLabels';
```
We're getting deep on the nesting here, I get that this is technically in the same parent folder. But do we want to use the @ notation here?

engbergandreas commented on src/components/Property/Types/StringProperty.tsx:
```diff
@@ -1,27 +1,28 @@
-  value
-}: Props) {
+export function StringProperty({ name, disabled, setPropertyValue, value }: Props) {
+  if (disabled) {
+    return (
+      <Paper px={'sm'} py={5}>
+        <Text size={'sm'}>{value}</Text>
+      </Paper>
+    );
+  }
   return (
     <StringInput
       disabled={disabled}
```
Why don't we want to use the string input if disabled? 
If we are to keep it this way we can remove line 22

engbergandreas commented on src/components/Property/Types/VectorProperty/ViewOptions/ColorVector.tsx:
```diff
@@ -0,0 +1,38 @@
+
+import { ValueList } from './DefaultValueList';
+
+export function ColorVector(props: VectorPropertyProps) {
+  const { disabled, setPropertyValue, value, isInt } = props;
```
```suggestion
export function ColorVector({disabled, setPropertyValue, value, isInt} : VectorPropertyProps) {
```
??

WeirdRubberDuck replied:
Unfortunately, no. This property type uses both the regular view for properties and for that one I need to spread all the inputted properties. And including the entire list here seemed less readable to me

Maybe it's just a bad design of the components. Would be open to discuss alternative solutions

ylvaselling replied:
Yeah let's try redesign at some point :)

engbergandreas commented on src/components/Property/Types/VectorProperty/ViewOptions/ColorVector.tsx:
```diff
@@ -0,0 +1,38 @@
+          b: Math.round(255 * value[2]),
+          a: hasAlpha ? value[3] : 1.0
+        }}
+        onChange={(rgbaColor: RGBA) => {
+          const newValue = [rgbaColor.r / 255, rgbaColor.g / 255, rgbaColor.b / 255];
+          if (hasAlpha) {
+            newValue.push(rgbaColor.a);
+          }
+          setPropertyValue(newValue);
+        }}
```
check for min/max values here?

engbergandreas reviewed (commented):
Looking solid!! 🎉 
A few general things: 

1. Read only property grows larger than other properties 
![image](https://github.com/user-attachments/assets/1f178cc1-c3b0-4932-bf81-1b6a451c1f76)

2. Getting a bunch of redux warnings when making changes to many of the properties, don't know how big of an issue this actually is though.. 
![image](https://github.com/user-attachments/assets/f58120d7-c85a-49d7-8bca-32ca9087bfaf)


3. The behaviour of sliders when either on the slider itself or using stepper controls are weird. Going down/back steps twice and forward/up nothing happens. Also, the focus highlight on the slider does not work in OpenSpace (browser is fine) 
https://github.com/user-attachments/assets/07ebb92d-7796-4191-8e08-4a3d70ba925f

4. We should have some limit on color inputs so that this is not valid, both picker and input: 
`rgb(30, 45, 2200)`

5. The text marker is not visible in input fields, here I'm positioned between 8 and 2
![image](https://github.com/user-attachments/assets/71e3ca21-4cb9-45f4-99cd-8ce715e84c67)

WeirdRubberDuck commented on src/components/ColorPicker/ColorEdit.tsx:
```diff
@@ -0,0 +1,88 @@
+            }
+          }
+        }}
+        onChange={setTextEditValue}
```
The input fields should be able to take any type of input, but we should add warnings

The color however, do need some validation. Correct

WeirdRubberDuck commented:
> Looking solid!! 🎉 A few general things:
> 
> 1. Read only property grows larger than other properties
>    ![image](https://private-user-images.githubusercontent.com/48772850/421850890-1f178cc1-c3b0-4932-bf81-1b6a451c1f76.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njc4OTIsIm5iZiI6MTc0MTg2NzU5MiwicGF0aCI6Ii80ODc3Mjg1MC80MjE4NTA4OTAtMWYxNzhjYzEtYzNiMC00OTMyLWJmODEtMWI2YTQ1MWMxZjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMDYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwOTVjZWE5ZTVlNWUzNWYzNDM5M2EzYzk2MmVmODUyOTc2NGU4NmExZWFlMWI5YTljZTFmMDBlYjI5MmQ3OTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6y-dE1OgFGz1CNt-vUtUQ1BBEYibo0uspSY1Jte2Seo)
> 2. Getting a bunch of redux warnings when making changes to many of the properties, don't know how big of an issue this actually is though..
>    ![image](https://private-user-images.githubusercontent.com/48772850/421851235-f58120d7-c85a-49d7-8bca-32ca9087bfaf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njc4OTIsIm5iZiI6MTc0MTg2NzU5MiwicGF0aCI6Ii80ODc3Mjg1MC80MjE4NTEyMzUtZjU4MTIwZDctYzg1YS00OWQ3LThiY2EtMzJjYTkwODdiZmFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMDYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlYWQ4MjY5MzBkYTc3ZGI4ODM4Yjc2NTEzYzg3MDQzOGIyODIwYTRiNzYwZWQ2ZWNjYTRhNTQxMjM5ZmZhZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FZRACbkEVlZjAI0KbCHN2HtacgRJPD-0n1V-kykRqoU)
> 3. The behaviour of sliders when either on the slider itself or using stepper controls are weird. Going down/back steps twice and forward/up nothing happens. Also, the focus highlight on the slider does not work in OpenSpace (browser is fine)
>    https://github.com/user-attachments/assets/07ebb92d-7796-4191-8e08-4a3d70ba925f
> 4. We should have some limit on color inputs so that this is not valid, both picker and input:
>    `rgb(30, 45, 2200)`
> 5. The text marker is not visible in input fields, here I'm positioned between 8 and 2
>    ![image](https://private-user-images.githubusercontent.com/48772850/421853556-71e3ca21-4cb9-45f4-99cd-8ce715e84c67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njc4OTIsIm5iZiI6MTc0MTg2NzU5MiwicGF0aCI6Ii80ODc3Mjg1MC80MjE4NTM1NTYtNzFlM2NhMjEtNGNiOS00NWY0LTk5Y2QtOGNlNzE1ZTg0YzY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMDYzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI5NTA2MTljMDZmMGY3MWFkY2U1ZWQxN2IyYjJlYzE2ZjkyOTY4MWIyMjU5Yzk1MTFkMTg0YjYyMjFlZmFiMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.N1045lL4e4DOtZPDwQYlWC2_HRtpEvYv0_bLdAYLnzM)

1. ✅ Is fixed
2. ⛔ Yeah, not sure what to do about this. Requires some investigation into the redux I think. Separate from this PR.
3. Not sure what you mean by this one - can't reproduce the behavior you're describing
4. Yes! TODO
5. ✅ It is, but you need to use the OpenSpace branch for the UI update. That indicator depends on the fixes for focus in that branch

WeirdRubberDuck commented:
> Overall I think it looks awesome!! ✨ I just jotted down some notes from looking at the properties. Lmk if you want to discuss
> 
> 1. General
>    
>    1. Maybe we can make the titles less bold? They are a little hard to read now imo.
>    2. I just had a thought - we could potientially go around the min max problem by making them as tiny input fields so you could set them yourself.
> 2. Numeric property
>    
>    1. Overlap of text over the textbox
>       ![image](https://private-user-images.githubusercontent.com/24265984/415591365-fe319eeb-524c-4362-b0b0-37d0ac8120e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTEzNjUtZmUzMTllZWItNTI0Yy00MzYyLWIwYjAtMzdkMGFjODEyMGUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1NDMxMDY0MGFiYTE2NTc0ZDA5Y2UwZDQ3ZGFmMTU1NzgxMzdhOTNjYjY5OGU0ZjM4NWU3ZDQyZjk5ZDI0NDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9nLw88E24nyuJWH1ti1KAMRdDirV79gH1FNu5FePhK0)
>    2. Is there a way to make these look more connected? When the window is small it is hard to see that these two are the same property.
>    3. Why does the slider have dots for these values? Seems 0.05 is always in there?
>    4. Should we somehow let the user know when they are out of range? Like we did in the old UI
>    5. Seems the circle slider doesn't go all the way to the right? I think there might be an offset to the blue bar? ![image](https://private-user-images.githubusercontent.com/24265984/415593908-bc19cc48-27aa-4592-b29e-d510c7e7592c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTM5MDgtYmMxOWNjNDgtMjdhYS00NTkyLWIyOWUtZDUxMGM3ZTc1OTJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhODg2ZGIzZjgwYWI5NTRjOWQ1MTYyZThkMTRiZDQ4YzYzM2JiN2NlMzljMWM3ZmI3OWU0MjcxOTZhMDk2MTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pLy-YL4VgQ-rCfUJV7DGVQpFFRT8HaoCwaFODnj6Urc)
>       ![image](https://private-user-images.githubusercontent.com/24265984/415600290-67022d3d-cc22-4a02-9134-df68e7c633d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU2MDAyOTAtNjcwMjJkM2QtY2MyMi00YTAyLTkxMzQtZGY2OGU3YzYzM2Q1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1NTZiMTU2ZjZjYWJkMGVjYzNkNDZlMzVmNTRiN2Q4MWM4MTA2YTU3NzQ2ZTVlYTAzOWQ2ZjhiNjhhY2I2YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uSdc6scQ3-KkV--Y-GQXOXXuHZVPGUuKJLCKS1X3QLA)
>    6. It seems when the property is line wrapped it goes outside bounds? ![image](https://private-user-images.githubusercontent.com/24265984/415595385-79a9278b-f890-407e-bd4b-5d61408806e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTUzODUtNzlhOTI3OGItZjg5MC00MDdlLWJkNGItNWQ2MTQwODgwNmU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyNDIzZGZlYTE4YzIxNTRhOTM3ZDY1NDc0MjIwY2FkYTIzZGRkNzQ0YTRkZTBhZTc0NDdkYzRjMTkyMDEwYTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.d-zoOWMB8eu-ij8NXnnHa6LduAvADk9YyZOxZlmrE8E)
> 3. String list property
>    Can we somehow make this smaller? Like a dropdown?
>    ![image](https://private-user-images.githubusercontent.com/24265984/415592207-2bbeba08-7243-4d9a-b25c-82bbb4e69da9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTIyMDctMmJiZWJhMDgtNzI0My00ZDlhLWIyNWMtODJiYmI0ZTY5ZGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhZTQwMGNiZDE2ZDE2ZmM4MWViMjM3NjA4Njg5NWNiMWVhMDU1NWM3MGVlNzMzYTYzZDk0ODFiMDM4ZGFmZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3dw3lIPZ9KHQpisnFvl1CuX0FCeIXNaAJTfPuu3sLLY)
> 4. Vec2 property
>    In the gui it took me a while to realize this was probably a vec2 property. I assume this is x and y? Can we maybe write that out somehow if that is the case? This might go for all vec properties - we might want to write out xyzw?
>    ![image](https://private-user-images.githubusercontent.com/24265984/415597404-71fec6b3-bd9c-4f66-8745-dab44413eafe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTc0MDQtNzFmZWM2YjMtYmQ5Yy00ZjY2LTg3NDUtZGFiNDQ0MTNlYWZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZGZmNzRmNzVjZTdhNTA5MzgwOTNhNzkwMDVhZTU2NTE3ZjVkZTAwNTRiZTM4NjM3YmFmZGNmYzFkNTBlZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9MGaxaNa7bg2wZ1Qo3Z0aucTBdH_-EVsyOuOoiZIRCo)
> 5. Selection property
>    Make a placeholder when empty?
> 6. Min max
>    Should we place the slider in the middle of the the number? Or the slider above the numbers, to mimic the numeric property when it is linewrapped?
> 7. Infobox
>    Can we place it so it is not in different places depending on how long the title is? Maybe to the far right?
> 8. Propertyowner
>    
>    1. Some padding on the right hand side? ![image](https://private-user-images.githubusercontent.com/24265984/415595766-53ae469c-7218-4d07-95d0-96e6fe8dd8d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTU3NjYtNTNhZTQ2OWMtNzIxOC00ZDA3LTk1ZDAtOTZlNmZlOGRkOGQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1YzBjNDVkZWVmYzRlMGYyZjlmM2ZiNDA2YWIyNTFhNjlmOWNkZGI2ZDY0ZTdjZTAwYjAzMjFhNmRkODg0MDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1S3rbomDylOvFYtGSjZPtGSXeZwQDZQX4RV081ALnF8)
>    2. Align about with the upper border of the about text?
>       ![image](https://private-user-images.githubusercontent.com/24265984/415596637-dadc5f01-1430-4b14-8480-0eff93ccc5db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTY2MzctZGFkYzVmMDEtMTQzMC00YjE0LTg0ODAtMGVmZjkzY2NjNWRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5NTdmMjU0NWE5YTM5MjhlYTdiMDUyZTM1Y2FiMjI2ZjQ0Njk2MjMxNjU0MGVkYzYyYjkwYzI3NmQ2MDMxMWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jRKCJoPOdoKmKTdw0lqAehDU8UOrbHqlzPcUsd36d5M)
> 9. Seems we have an issue with rendering
>    ![image](https://private-user-images.githubusercontent.com/24265984/415606183-d4e27fc5-51c7-484d-bab1-93f24544ec16.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU2MDYxODMtZDRlMjdmYzUtNTFjNy00ODRkLWJhYjEtOTNmMjQ1NDRlYzE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5MjZkMGJkN2U4Yzk0NDJjYjNmOGUwYzY2ZGZmOWRmZWM2OWZhOWJlY2IzMDU2MGQyMDlkYzA1YTdhYmU4NzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xigewG1d78lxqrMLTkNT7DUAjl42V9LnvzTGUtB4Al8)
> 10. Wth 😆 We really like saying Earth is a special place
>     ![image](https://private-user-images.githubusercontent.com/24265984/415596305-a028d1f9-7ce6-465e-be0a-a4bbf2075f8f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4Njk0MzgsIm5iZiI6MTc0MTg2OTEzOCwicGF0aCI6Ii8yNDI2NTk4NC80MTU1OTYzMDUtYTAyOGQxZjktN2NlNi00NjVlLWJlMGEtYTRiYmYyMDc1ZjhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEzVDEyMzIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3N2VlMzhhYzZkMGNiZjUzODYwZTI5MTY0Y2Q1YWQ4YzRmMWRhYjU5YzFhYzRmNzNiMTY3ZTMxZGU2MWI3YmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NtKX7w2cvJx4IGezPEQD24TpfcLo6_Roza5gI1KlcXo)

1. ✅ For now I changed the font weight of the input fields. I like the idea of min and max as input values, but we should save that for a future design iteration (maybe once we can set the min and max per object, even :) would be quite natural then) 
2. ✅ Essentially all of these points should be fixed. 
  - One remaining TODO is to provide error messages when the input value is out of range, like we did in the old UI. 
3. ✅ Gave this property type a max height and added a scrollbox
4. TODO: Improve vector properties
5. ✅ Done
6. TODO: Think about this and test out an alternative look
7. TODO: Discuss 
8.  i) ✅ Done ii) TODO
9. ✅ I don't remember what the cause for this was, but I fixed it :) 
10. ✅ Haha, indeed... We need to go through our descriptions ;)

ylvaselling commented on src/components/ColorPicker/ColorEdit.tsx:
```diff
@@ -0,0 +1,86 @@
+  const formats = withAlpha ? ['rgba', 'hexa', 'hsla'] : ['rgb', 'hex', 'hsl'];
+  const defaultFormat = withAlpha ? 'rgba' : 'rgb';
+
+  const [format, setFormat] = useState<ColorFormat | undefined>(defaultFormat);
```
I don't think this can ever be undefined

ylvaselling commented on src/components/ColorPicker/ColorEdit.tsx:
```diff
@@ -0,0 +1,86 @@
+      />
+      <ColorInput
+        value={textEditValue}
+        withEyeDropper={false}
```
I think we should enable this

WeirdRubberDuck replied:
I would agree, and wondered why I disabled it.  It appears that unfortunately, it does not work in OpenSpace. And even if it did work, it would only work to pick colors form the UI itself, most likely. 

So I'd opt for removing it in all our color inputs

ylvaselling commented on src/components/ColorPicker/ColorEdit.tsx:
```diff
@@ -0,0 +1,86 @@
+        format={withAlpha ? 'rgba' : 'rgb'}
+        onChange={onColorChange}
+      />
+      <ColorInput
```
Needs some way to focus on this popup with keyboard navigation

WeirdRubberDuck replied:
Yep, noticed this yesterday too. As well as closing

ylvaselling commented on src/components/Property/Property.tsx:
```diff
@@ -71,32 +72,38 @@ interface Props {
   }
 
-  // console.log(property);
+  // In most cases we want to show the label on top of the property input, but some types
+  // have a custom label built in in the input type. Keep track of which that is
+  const hasBuiltInLabel =
+    ConcreteProperty === BoolProperty || ConcreteProperty === TriggerProperty;
```
These two properties have a built in label?

WeirdRubberDuck replied:
Yes, here we don't want to show the property label on top of the input, the same way we do for the others.  Coult try to clarify that in the comment?

ylvaselling replied:
Aha I see now what you mean! I think the comment is fine actually now that i now what you mean

ylvaselling commented on src/components/Property/PropertyLabel.tsx:
```diff
@@ -1,16 +1,32 @@
+            label={`This property is read-only, meaning that it's not intended to be changed.`}
+          >
+            <Text span ml={'xs'} size={'xs'} c={'dimmed'}>
+              (Read-only)
```
How about an icon for this? 
![Uploading image.png…]()

ylvaselling commented on src/components/Property/SliderUtil/hook.ts:
```diff
@@ -0,0 +1,40 @@
+  // When no min/max is set, the marks for the slider cannot be nicely computed
+  const marks = !isFinite(extent)
+    ? []
+    : [
+        {
+          value: min
+        },
+        { value: min + 1 * marksStep },
+        {
+          value: min + extent / 2
+        },
+        { value: min + 3 * marksStep },
+        {
+          value: max
+        }
+      ];
```
```suggestion
    :  [0, 1, 2, 3, 4].map((value, i) => { return { value: min + value * marksStep }});
```
Also this first array doesn't need the object with value in it:

```suggestion
    :  [0, 1, 2, 3, 4].map((value, i) => min + value * marksStep );
```

WeirdRubberDuck replied:
Good catch!

ylvaselling commented on src/components/Property/SliderUtil/hook.ts:
```diff
@@ -0,0 +1,40 @@
+        }
+      ];
+
+  const scaledMarks = marks.map((mark) => ({
+    value: scale.invert(mark.value)
+  }));
```
```suggestion
  const scaledMarks = marks.map((mark) => ({
    value: scale.invert(mark)
  }));
```
This needs a comment - why is the scale inverting the mark?

ylvaselling commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,153 @@
+  value,
+  valueType = 'string'
+}: Props) {
+  const [placeholderText, setPlaceholderText] = useState('');
+
+  // The input string that the user is typing
+  const [inputString, setInputString] = useState('');
+
+  // The values to shown as a list of pills in the input
+  const {
+    value: shownValues,
+    setValue: setShownValues,
+    setIsEditing,
+    isEditing
+  } = usePropListeningState(value);
+
+  const [clickedItemIndex, setClickedItemIndex] = useState<number | undefined>(undefined);
```
Group the useState hooks together

ylvaselling commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,153 @@
+
+  const [clickedItemIndex, setClickedItemIndex] = useState<number | undefined>(undefined);
+
+  const placeHolderText = useMemo(() => {
```
This value is shadowing the state

ylvaselling commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,153 @@
+      <Pill.Group mah={100} style={{ overflowY: 'auto' }}>
+        {shownValues.map((v, i) => (
+          <Pill
+            key={`pill-${i}`}
```
It is not advised to use index as the key

WeirdRubberDuck replied:
Vell aware.  Not sure there is an alternative here, though?

ylvaselling commented on src/components/Property/Types/ListProperty/ListProperty.tsx:
```diff
@@ -0,0 +1,153 @@
+      aria-label={`List input for ${name}`}
+    >
+      <Pill.Group mah={100} style={{ overflowY: 'auto' }}>
+        {shownValues.map((v, i) => (
```
```suggestion
        {shownValues.map((value, i) => (
```

ylvaselling commented on src/components/Property/Types/NumericProperty/NumericProperty.tsx:
```diff
@@ -0,0 +1,100 @@
+        ) : (
+          <NumericInput
+            value={currentValue}
+            valueLabel={(v) =>
```
```suggestion
            valueLabel={(value) =>
```

ylvaselling commented on src/components/Property/Types/NumericProperty/Slider/NumericPropertySlider.tsx:
```diff
@@ -0,0 +1,72 @@
+  return (
+    <Stack gap={0} {...styleProps}>
+      <Slider
+        label={(v) => roundNumberToDecimalPlaces(v, decimalPlaces)}
```
```suggestion
        label={(value) => roundNumberToDecimalPlaces(value, decimalPlaces)}
```

ylvaselling replied:
Same for the rest of the code... I think we decided we weren't going to have one letter variables unless they were indices

ylvaselling commented on src/components/Property/Types/NumericProperty/util.ts:
```diff
@@ -0,0 +1,9 @@
+export function stepToDecimalPlaces(step: number): number {
+  const stepStr = step.toString();
+  const decimalPlaces = stepStr.includes('.') ? stepStr.split('.')[1].length : 0;
```
Are you getting the last element of the array with the 1 index? how about `pop()` ?
```suggestion
  const decimalPlaces = stepStr.includes('.') ? stepStr.split('.')[1].length : 0;
```

ylvaselling commented on src/components/Property/Types/VectorProperty/ViewOptions/ColorVector.tsx:
```diff
@@ -0,0 +1,38 @@
+export function ColorVector(props: VectorPropertyProps) {
+  const { disabled, setPropertyValue, value, isInt } = props;
+
+  if (value.length < 3 || value.length > 4 || isInt) {
```
```suggestion
  if (value.length !== 3 || value.length !== 4 || isInt) {
```

WeirdRubberDuck replied:
It would have to look like this: 
```
 if ((value.length !== 3 && value.length !== 4) || isInt) {
```
but sure
